numpyro/contrib/module.py

```diff
@@ -71,6 +82,150 @@ def haiku_module(name, nn, input_shape=None):
             raise ValueError('Valid value for `input_shape` needed to initialize.')
         # feed in dummy data to init params
         rng_key = numpyro.sample(name + '$rng_key', PRNGIdentity())
-        nn_params = nn.init(rng_key, jnp.ones(input_shape))
-        numpyro.param(module_key, nn_params)
-    return partial(nn.apply, nn_params, None)
+        nn_params = nn_module.init(rng_key, jnp.ones(input_shape))
+        # haiku init returns an immutable dict
+        # we cast it to a mutable one to be able to set priors for parameters
+        nn_params = numpyro.param(module_key, haiku.data_structures.to_mutable_dict(nn_params))
+    return partial(nn_module.apply, nn_params, None)
+
+
+# register an "empty" parameter which only stores its shape
+# so that the optimizer can skip optimize this parameter, while
+# it still provides shape information for priors
+ParamShape = namedtuple("ParamShape", ["shape"])
+register_pytree_node(ParamShape, lambda x: ((None,), x.shape), lambda shape, x: ParamShape(shape))
+
+
+def _update_params(params, new_params, prior, prefix=''):
+    """
+    A helper to recursively set prior to new_params.
+    """
+    for name, item in params.items():
+        flatten_name = ".".join([prefix, name]) if prefix else name
+        if isinstance(item, dict):
+            assert not isinstance(prior, dict) or flatten_name not in prior
+            new_item = new_params[name]
+            _update_params(item, new_item, prior, prefix=flatten_name)
+        elif (not isinstance(prior, dict)) or flatten_name in prior:
+            d = prior[flatten_name] if isinstance(prior, dict) else prior
+            if isinstance(params[name], ParamShape):
+                param_shape = params[name].shape
+            else:
+                param_shape = jnp.shape(params[name])
+                params[name] = ParamShape(param_shape)
+            param_batch_shape = param_shape[:len(param_shape) - d.event_dim]
+            # XXX: here we set all dimensions of prior to event dimensions.
+            new_params[name] = numpyro.sample(flatten_name, d.expand(param_batch_shape).to_event())
+
+
+def random_flax_module(name, nn_module, prior, *, input_shape=None):
+    """
+    A primitive to place a prior over the parameters of the Flax module `nn_module`.
+
+    :param str name: name of NumPyro module
+    :param flax.nn.Module: the module to be registered with NumPyro
+    :param prior: a NumPyro distribution or a Python dict with parameter names as keys and
+        respective distributions as values. For example::
+
+            net = random_flax_module("net",
```

Comment:
How are we aligning the names in the `"prior"` dict and the NN parameter names? Are they easily available for a flax model, and what happens to the parameter names when multiple modules are combined?

Reply:
Parameters of `flax` and `haiku` are stored in nested dicts. For example, given the below model:
```
{'Dense_0': {'kernel': DeviceArray([[-0.5536316,  1.1597457, -0.6334757]], dtype=float32),
  'bias': DeviceArray([0., 0., 0.], dtype=float32)},
 'Dense_1': {'kernel': DeviceArray([[-0.09242683, -0.05430904, -0.3762646 ],
               [-0.1693897 ,  0.14864841, -0.57235295],
               [ 0.19652358,  0.6845585 , -0.71639156]], dtype=float32),
  'bias': DeviceArray([0., 0., 0.], dtype=float32)},
 'Dense_2': {'kernel': DeviceArray([[ 0.38516834],
               [-0.25442997],
               [-0.8081337 ]], dtype=float32),
  'bias': DeviceArray([0.], dtype=float32)},
 'Dense_3': {'kernel': DeviceArray([[-0.46138018],
               [ 0.3994321 ],
               [-0.8642299 ]], dtype=float32),
  'bias': DeviceArray([0.], dtype=float32)}}
```
An example for prior is: `{"Dense_1.kernel": dist.Normal(0, 1), "Dense_3.bias": dist.Normal(0, 1)}`. I used **flat** name convention as in PyTorch for this, as in [test_update_params](https://github.com/pyro-ppl/numpyro/pull/743/files#diff-b710b48cf627f2b6524be99a2d4194dbR60). Alternatively, we can use nested dict for `prior` here but I found it less flexible. What do you think?

Reply:
If we decide to go with the "flatten name" convention, then I will need to add a note for it.

Reply:
I think a flattened name convention (over nested dictionary) is fine. It would be more user-friendly if users could specify the prior at the level of the module itself so that they didn't need to understand the naming convention for chained Flax modules, but that will require a `NumpyroModule` that inherits from a Flax module and can be chained together similarly. I'll think about it. Lets proceed with this design for the time being.